### PR TITLE
[Rust][IRModule] Flesh out IRModule methods

### PR DESCRIPTION
**Pull request by jroesch**
This PR fully expand the IRModule interface in Rust and fixes a whole host of sharp edges and improves the procedural macros for generating the boilerplate. 

cc @mwillsey @altanh @imalsogreg @gussmith23

### CODE REVIEW

**Review by junrushao1994 — COMMENTED** [review]
The LLVM flags look good to me

File: rust/tvm-macros/src/external.rs
Comment (imalsogreg) [inline rust/tvm-macros/src/external.rs:24]: nit: line length

File: rust/tvm-rt/src/value.rs
Comment (imalsogreg) [inline rust/tvm-rt/src/value.rs:25]: stray comment

File: rust/tvm-macros/src/external.rs
Comment (imalsogreg) [inline rust/tvm-macros/src/external.rs:26]: nit: brief rustdoc? Some comments would help me get the story of what `ExternalItem` and `External` are for.

File: rust/tvm-macros/src/object.rs
Comment (imalsogreg) [inline rust/tvm-macros/src/object.rs:194]: Possibly dumb comment (I'm not familiar with this macro) - is `self.0` always the right thing to hash? Or is `#ref_id` sometimes a tuple with multiple fields, where field `.1` and `.2` etc. may be part of what distinguishes objects when hashing?

File: rust/tvm-sys/src/datatype.rs
Comment (imalsogreg) [inline rust/tvm-sys/src/datatype.rs:86]: if this is `float32`, should the old `float` be renamed to `float64` (to avoid clashing with `float` vs. `double` size intuition?)

File: rust/tvm/src/ir/module.rs
Comment (imalsogreg) [inline rust/tvm/src/ir/module.rs:150]: Is `clone()` on an `IRModule` expensive?

**Review by imalsogreg — APPROVED** [review]
I left a couple nit comments and questions. But it seems fine to merge, to me.
Haven't tested anything (I'm not up to speed yet on running TVM or using these rust bindings)

File: rust/tvm-macros/src/external.rs
Comment (adelbertc) [reply rust/tvm-macros/src/external.rs:24]: WTB rustfmt

Comment (jroesch) [reply rust/tvm-macros/src/external.rs:24]: I probably just haven't run it yet, since everything is automatic I tend to just run it when I know I'm gtg.

File: rust/tvm-rt/src/value.rs
Comment (jroesch) [reply rust/tvm-rt/src/value.rs:25]: it has been purged

File: rust/tvm-sys/src/datatype.rs
Comment (jroesch) [reply rust/tvm-sys/src/datatype.rs:86]: We don't use float/double terminology in TVM due to the fact that both integers and floating point values are both parametric by vector length and bit width. The float method here takes both bit width and vector length, unless I'm missing something.

File: rust/tvm/src/ir/module.rs
Comment (jroesch) [reply rust/tvm/src/ir/module.rs:150]: No all the "language" types are a newtype around an Option<ObjectPtr<T>> and ObjectPtr<T> is roughly Arc<T> plus inheritance support.

**Comment by jroesch** [comment]
> I left a couple nit comments and questions. But it seems fine to merge, to me.
> Haven't tested anything (I'm not up to speed yet on running TVM or using these rust bindings)

Its pretty easy to get started, @adelbertc has not repro'd running them if you are interested. I will finish adding tests and ping you guys again.

File: rust/tvm-macros/src/object.rs
Comment (jroesch) [reply rust/tvm-macros/src/object.rs:202]: No we generate a wrapper new type over the ObjectPtr<T> which contains 1 field always. ObjectPtr is currently dispatching to TVM's C++ hashing so we have cross language consistency.

File: rust/tvm-macros/src/external.rs
Comment (adelbertc) [reply rust/tvm-macros/src/external.rs:29]: What Greg said

File: rust/tvm-macros/src/external.rs
Comment (adelbertc) [inline rust/tvm-macros/src/external.rs:154]: Is `abort!` the standard way to blow up a macro?

File: rust/tvm-macros/src/object.rs
Comment (adelbertc) [inline rust/tvm-macros/src/object.rs:82]: This is like if you mark something with a `no_derive` attribute you don't derive `Debug`? Naming is a bit wonky here

File: rust/tvm-rt/src/array.rs
Comment (adelbertc) [inline rust/tvm-rt/src/array.rs:87]: Instead of consuming `self` with `into_iter` could you have a method that returns an `Iterator` over references so you don't have to clone, since `write!` I assume doesn't need owned values? But this may not matter if `clone`ing these `Array`s isn't expected to be expensive.

File: rust/tvm-rt/src/object/object_ptr.rs
Comment (adelbertc) [inline rust/tvm-rt/src/object/object_ptr.rs:151]: remove dead code

**Review by adelbertc — APPROVED** [review]

File: rust/tvm-macros/src/external.rs
Comment (jroesch) [reply rust/tvm-macros/src/external.rs:154]: This is proc_macro_error which is crate for doing better errors in proc macros.

File: rust/tvm-macros/src/object.rs
Comment (jroesch) [reply rust/tvm-macros/src/object.rs:82]: yeah currently I just turn off all deriving for types marked no-derive (this all to support String which is like complete bespoke impl)

File: rust/tvm-rt/src/array.rs
Comment (jroesch) [reply rust/tvm-rt/src/array.rs:87]: its just cheaper to clone the outer smart pointer because we only modify one reference count instead of touching the ref count on every single element.

File: rust/tvm-rt/src/object/object_ptr.rs
Comment (jroesch) [reply rust/tvm-rt/src/object/object_ptr.rs:151]: I was leaving this here for after I land another outstanding branch so I can clean up some of the code.

**Comment by jroesch** [comment]
@adelbertc @imalsogreg I will try and ship docs for proc macro after landing this branch if that's cool with you guys, just so I can unblock some other branches I have.

**Review by tqchen — APPROVED** [review]

**Review by electriclilies — APPROVED** [review]
I don't know rust very well, but looks good to me

File: rust/tvm-rt/src/array.rs
Comment (mdwelsh) [inline rust/tvm-rt/src/array.rs:130]: Inconsistency: `FromIterator::from_iter()` is implemented elsewhere using
```
Array::from_vec(iter)
```
See the following similar code:
* array.rs:92
* graph_mod.rs:139
* graph_mod.rs:151

For more information about this comment, see https://fixie.ai/comment/55732b1c0.

File: rust/tvm-rt/src/object/object_ptr.rs
Comment (mdwelsh) [inline rust/tvm-rt/src/object/object_ptr.rs:162]: Remove dead code.

For more information on this comment, see https://fixie.ai/comment/68314539a.

Comment (mdwelsh) [reply rust/tvm-rt/src/object/object_ptr.rs:162]: I'm leaving the code in for now, since I plan to refactor this later.

Comment (mdwelsh) [reply rust/tvm-rt/src/object/object_ptr.rs:162]: OK, resolving.

**Pull request merged**
